### PR TITLE
create a recipe template

### PR DESCRIPTION
**Pull request by ericLemanissier**
this can help new contributors, and reviewers too.

This is a very quick draft.
We should probably also add more comments to this template, in order to explain all these conventions.
Also, if you find this template could be useful, then we should link to it from the documentation .md files

I added all CCI's approved reviewers as contributor to my fork, so please edit the files directly, instead of doing code suggestion.

this partially answers item 3. in https://github.com/conan-io/conan-center-index/issues/2857#issuecomment-860785693

### CODE REVIEW

**Comment by danimtb** [comment]
It would be very nice if combined with the `conan new --template` command https://docs.conan.io/en/latest/extending/template_system/command_new.html#template-command-new

**Comment by ericLemanissier** [comment]
@SpaceIm maybe you have some sort of checklist you use for modernization of recipes ? Can it somehow be translated into this template ?

**Review by intelligide — APPROVED** [review]

**Comment by SSE4** [comment]
I also had some templates, but didn't update them for a while: https://github.com/SSE4/cci.templates

**Review by SSE4 — APPROVED** [review]

**Comment by ericLemanissier** [comment]
I'll do a pr on your repo then. It's more usable as a proper template. I did not know not know that `conan new`could use templates. Can we put a link in a doc to your template repo? Can these templates be put in CCI?

File: docs/package/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    generators = "cmake", "cmake_find_package_multi"
+
+    # no manual caching of build helper like  _cmake = None
+
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def _build_subfolder(self):
```
Comment (StellaSmith) [inline docs/package/all/conanfile.py:34]: Shouldn't this be a property?

Comment (ericLemanissier) [reply docs/package/all/conanfile.py:34]: Yes!

File: docs/package/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+from conan.tools.microsoft import msvc_runtime_flag
+import functools
+import os
+
+required_conan_version = ">=1.36.0"
+
+class packageConan(ConanFile):
+    name = "package"
+    description = "shortd escription"
+    license = "" # conform to SPDX License List
```
Comment (uilianries) [inline docs/package/all/conanfile.py:12]:
```suggestion
    license = "" # conform to SPDX License List: https://spdx.org/licenses/
```

File: docs/package/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/project/package"
+    topics = ("topic1", "topic2") # no "conan" in topics
```
Comment (uilianries) [inline docs/package/all/conanfile.py:15]:
```suggestion
    topics = ("topic1", "topic2") # no "conan"  and project name in topics
```

File: docs/package/all/conanfile.py (hunk 1 above)
Comment (uilianries) [inline docs/package/all/conanfile.py:34]:
```suggestion
    @property
    def _build_subfolder(self):
```

File: docs/package/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def _build_subfolder(self):
+        return "build_subfolder"
+
+    @property
+    def _is_msvc(self):
```
Comment (uilianries) [inline docs/package/all/conanfile.py:38]: On Conan 1.45 there is a native version: https://docs.conan.io/en/latest/reference/conanfile/tools/microsoft.html#conan-tools-microsoft-is-msvc

Soon, 1.45.0 will be available on Conan Center Index.

**Review by uilianries — COMMENTED** [review]
Excellent initiative!

**Comment by SSE4** [comment]
> I'll do a pr on your repo then. It's more usable as a proper template. I did not know not know that `conan new`could use templates. Can we put a link in a doc to your template repo? Can these templates be put in CCI?

yes, why not. I have been thinking how to properly organize that, but probably forgot about it at some point in time.

File: docs/package/all/conanfile.py (hunk 4 above)
Comment (jgsogo) [reply docs/package/all/conanfile.py:38]: Now it is.

File: docs/package/all/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,7 @@
+cmake_minimum_required(VERSION 3.1)
+project(cmake_wrapper C)
+
+include(conanbuildinfo.cmake)
+conan_basic_setup(TARGETS KEEP_RPATHS)
+
+add_subdirectory(source_folder)
```
Comment (StellaSmith) [inline docs/package/all/CMakeLists.txt:7]:
```suggestion
add_subdirectory(source_subfolder)
```

**Review by StellaSmith — CHANGES_REQUESTED** [review]

File: docs/package/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        tools.remove_files_by_mask(os.path.join(self.package_folder, "lib"), "*.pdb")
+
+    def package_info(self):
+        self.cpp_info.libs = ["package_lib"]
+
+        self.cpp_info.set_property("cmake_find_mode", "both")
+        self.cpp_info.set_property("cmake_module_file_name", "PACKAGE")
+        self.cpp_info.set_property("cmake_file_name", "package")
+        self.cpp_info.set_property("cmake_target_name", "PACKAGE::PACKAGE")
+        self.cpp_info.set_property("pkg_config_name", "package")
```
Comment (SpaceIm) [inline docs/package/all/conanfile.py:120]:
```suggestion
        # if package is listed in https://cmake.org/cmake/help/latest/manual/cmake-modules.7.html#find-modules
        self.cpp_info.set_property("cmake_find_mode", "both")
        self.cpp_info.set_property("cmake_module_file_name", "PACKAGE")
        self.cpp_info.set_property("cmake_module_target_name", "PACKAGE::PACKAGE")
        # if package provides a config file (package-config.cmake)
        self.cpp_info.set_property("cmake_file_name", "package")
        self.cpp_info.set_property("cmake_target_name", "package::package")
        # if package provides a pkgconfig file (package.pc)
        self.cpp_info.set_property("pkg_config_name", "package")
```

Comment (SpaceIm) [inline docs/package/all/conanfile.py:120]:
```suggestion
        # if package is listed in https://cmake.org/cmake/help/latest/manual/cmake-modules.7.html#find-modules
        self.cpp_info.set_property("cmake_find_mode", "both")
        self.cpp_info.set_property("cmake_module_file_name", "PACKAGE")
        self.cpp_info.set_property("cmake_module_target_name", "PACKAGE::PACKAGE")
        # if package provides a config file (package-config.cmake)
        self.cpp_info.set_property("cmake_file_name", "package")
        self.cpp_info.set_property("cmake_target_name", "package::package")
        # if package provides a pkgconfig file (package.pc)
        self.cpp_info.set_property("pkg_config_name", "package")
```

File: docs/package/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        if self.settings.os in ["Linux", "FreeBSD"]:
+            self.cpp_info.system_libs.append("m")
+            self.cpp_info.system_libs.append("pthread")
+            self.cpp_info.system_libs.append("dl")
+
+
+
+        #  TODO: to remove in conan v2 once cmake_find_package_* generators removed
+        self.cpp_info.filenames["cmake_find_package"] = "PACKAGE"
+        self.cpp_info.filenames["cmake_find_package_multi"] = "PACKAGE"
```
Comment (SpaceIm) [inline docs/package/all/conanfile.py:132]:
```suggestion
        self.cpp_info.filenames["cmake_find_package_multi"] = "package"
```
```suggestion
        self.cpp_info.filenames["cmake_find_package_multi"] = "PACKAGE"
```

Comment (SpaceIm) [inline docs/package/all/conanfile.py:132]:
```suggestion
        self.cpp_info.filenames["cmake_find_package_multi"] = "package"
```
```suggestion
        self.cpp_info.filenames["cmake_find_package_multi"] = "PACKAGE"
```

File: docs/package/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        self.cpp_info.names["cmake_find_package"] = "PACKAGE"
+        self.cpp_info.names["cmake_find_package_multi"] = "PACKAGE"
```
Comment (SpaceIm) [inline docs/package/all/conanfile.py:134]:
```suggestion
        self.cpp_info.names["cmake_find_package_multi"] = "package"
```
```suggestion
        self.cpp_info.names["cmake_find_package_multi"] = "PACKAGE"
```

Comment (SpaceIm) [inline docs/package/all/conanfile.py:134]:
```suggestion
        self.cpp_info.names["cmake_find_package_multi"] = "package"
```
```suggestion
        self.cpp_info.names["cmake_find_package_multi"] = "PACKAGE"
```

File: docs/package/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    settings = "os", "arch", "compiler", "build_type" # even for header only
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+    }
+    generators = "cmake", "cmake_find_package_multi"
```
Comment (SpaceIm) [inline docs/package/all/conanfile.py:25]:
```suggestion

    # cmake generator as a bare minimum
    # cmake_find_package if at least one dependency is listed in https://cmake.org/cmake/help/latest/manual/cmake-modules.7.html#find-modules
    # cmake_find_package_multi for other dependencies
    # pkg_config if upstream CMakeLists uses pkg_check_modules() (don't forget to also add pkgconf in build requirements)
    generators = "cmake", "cmake_find_package_multi"
```

Comment (SpaceIm) [inline docs/package/all/conanfile.py:25]:
```suggestion

    # cmake generator as a bare minimum
    # cmake_find_package if at least one dependency is listed in https://cmake.org/cmake/help/latest/manual/cmake-modules.7.html#find-modules
    # cmake_find_package_multi for other dependencies
    # pkg_config if upstream CMakeLists uses pkg_check_modules() (don't forget to also add pkgconf in build requirements)
    generators = "cmake", "cmake_find_package_multi"
```

File: docs/package/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+    def requirements(self):
+        self.requires("dependency/0.8.1")
+
+    def validate(self):
+        if self.settings.compiler.get_safe("cppstd"):
+            tools.check_min_cppstd(self, 11)
+        if is_msvc(self) and self.options.shared:
+            raise ConanInvalidConfiguration("package can't be built as shared on visual studio")
+
```
Comment (SpaceIm) [inline docs/package/all/conanfile.py:73]:
```suggestion

    # if another tool than the compiler or CMake is required to build the project (pkgconf, bison, flex etc)
    def build_requirements(self):
        self.requires("tool/x.y.z")

```

Comment (SpaceIm) [inline docs/package/all/conanfile.py:73]:
```suggestion

    # if another tool than the compiler or CMake is required to build the project (pkgconf, bison, flex etc)
    def build_requirements(self):
        self.requires("tool/x.y.z")

```

File: docs/package/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+from conans.errors import ConanInvalidConfiguration
+from conan.tools.microsoft import msvc_runtime_flag, is_msvc
+import functools
+import os
+
+required_conan_version = ">=1.45.0"
+
+class packageConan(ConanFile):
+    name = "package"
+    description = "shortd escription"
```
Comment (SpaceIm) [inline docs/package/all/conanfile.py:11]:
```suggestion
    description = "short description"
```

File: docs/package/all/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
+from conans import CMake, ConanFile, tools
+from conans.errors import ConanInvalidConfiguration
+from conan.tools.microsoft import msvc_runtime_flag, is_msvc
+import functools
+import os
+
+required_conan_version = ">=1.45.0"
+
+class packageConan(ConanFile):
```
Comment (SpaceIm) [inline docs/package/all/conanfile.py:9]:
```suggestion
class PackageConan(ConanFile):
```

File: docs/package/all/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+
+    def package(self):
+        self.copy(pattern="LICENSE", dst="licenses", src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
+
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+        tools.rmdir(os.path.join(self.package_folder, "share"))
+        tools.remove_files_by_mask(os.path.join(self.package_folder, "lib"), "*.pdb")
```
Comment (StellaSmith) [inline docs/package/all/conanfile.py:111]:
```suggestion
        tools.remove_files_by_mask(os.path.join(self.package_folder, "bin"), "*.pdb")
```

**Review by StellaSmith — CHANGES_REQUESTED** [review]

File: docs/package/all/conanfile.py (hunk 13 above)
Comment (ericLemanissier) [inline docs/package/all/conanfile.py:111]:
```suggestion
        tools.remove_files_by_mask(os.path.join(self.package_folder, "lib"), "*.pdb")
        tools.remove_files_by_mask(os.path.join(self.package_folder, "bin"), "*.pdb")
```

File: docs/package/all/conanfile.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+            tools.patch(**patch)
+        # remove bundled xxhash
+        tools.remove_files_by_mask(os.path.join(self._source_subfolder, "lib"), "whateer.*")
+        tools.replace_in_file(
+            os.path.join(self._cmakelists_subfolder, "CMakeLists.txt"),
+            "...",
+            "",
+        )
+
+    @functools.lru_cache(1)
```
Comment (MartinDelille) [inline docs/package/all/conanfile.py:98]: What does that mean ?

Comment (ericLemanissier) [reply docs/package/all/conanfile.py:98]: it's a [python thing](https://docs.python.org/3/library/functools.html#functools.lru_cache): when calling this function, python actually does the call once, and caches the resulting object. When the same function is called after that, the cached result is returned.

**Review by danimtb — APPROVED** [review]
Let's try to move this forward! At least this seems like a good starting point

**Review by SSE4 — APPROVED** [review]

**Pull request merged**